src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+    /// results.
+    ///
+    /// If we actually stored a set of every possible value, this could by done by passing
+    /// each datup to the function one-by-one and unioning the resulting sets. This is possible
+    /// when our values set is empty or contains a single datum, but when it contains a range,
+    /// we need a little metadata about the function to be able to infer a useful result.
+    fn flat_map(
+        self,
+        is_monotonic: Monotonic,
+        mut datum_map: impl FnMut(Datum<'a>) -> ResultSpec<'a>,
+    ) -> ResultSpec<'a> {
+        let null_spec = if self.nullable {
+            datum_map(Datum::Null)
+        } else {
+            ResultSpec::nothing()
+        };
+
+        let error_spec = ResultSpec {
+            fallible: self.fallible,
+            ..Self::nothing()
+        };
+
+        let values_spec = match self.values {
+            Values::Empty => ResultSpec::nothing(),
+            // If this range contains a single datum, just call the function.
+            Values::Within(min, max) if min == max => datum_map(min),
+            // If this is a range of booleans, we know all the values... just try them.
+            Values::Within(Datum::False, Datum::True) => {
+                datum_map(Datum::False).union(datum_map(Datum::True))
+            }
+            // Otherwise, if our function is monotonic, we can try mapping the input
+            // range to an output range.
+            Values::Within(min, max) => match is_monotonic {
+                Monotonic::Yes => {
+                    let min_column = datum_map(min);
+                    let max_column = datum_map(max);
+                    if min_column.nullable
+                        || min_column.fallible
+                        || max_column.nullable
+                        || max_column.fallible
+                    {
+                        ResultSpec::anything()
+                    } else {
+                        min_column.union(max_column)
+                    }
+                }
+                Monotonic::Maybe | Monotonic::No => ResultSpec::anything(),
+            },
+            Values::Nested(_) | Values::All => ResultSpec::anything(),
+        };
+
+        null_spec.union(error_spec).union(values_spec)
+    }
+}
+
+/// [Abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation) for
+/// [MirScalarExpr].
+///
+/// [MirScalarExpr::eval] implements a "concrete interpreter" for the expression type: given an
+/// expression and specific column values as input, it returns a specific value for the output.
+/// This could be reimplemented using this trait... but it's most useful for "abstract"
+/// interpretations of the expression, where we generalize about sets of possible inputs and outputs.
+/// See [Trace] and [ColumnSpecs] for how this can be useful in practice.
+pub trait Interpreter {
+    type Summary: Sized + Clone;
```

Comment:
Any reason to not require `Display` or `Debug`?

Reply:
Not really! I can see a `Debug` bound making it easier to trace through the logic; I'll add it in.